### PR TITLE
JAX: Setup D2L JAX API & Add chapter linear regression 

### PR DESCRIPTION
**Pull request by AnirudhDagar**
This PR sets up the base for rest of the `JAX` port in D2L. The review may not be straight fwd, I'll share a few pointers below that went into the decision making for the design:

1. We decided to use `flax` and `optax` with `JAX` given they are more popular in the community compared to their counterparts.
2. `JAX` is meant to be written in a very functional form, but this is actually the exact opposite of how D2L V1.0 API works. The d2l api relies on OOP which may not be the natural way for writing pure `JAX` but on the up side, for a beginner once they grab the Lightning like D2L API, things get a lot easier.
3. Flax uses Python DataClasses (introduced in py3.7) which help reduce the boilerplate required to instantiate any class with several attribute fields. Since this is fairly new feature, it is probably not very widely adopted. DataClasses are harder to work with when dealing with inheritance (though this should be much better once we switch to py3.10, then we can utilize `kw_only` attr for the data fields)
4. Since JAX arrays are immutable and inplace operations are prohibited (can still do that with `.at`) but JAX ecosystem avoids inplace ops, this can lead to code being slightly more verbose in some places. It is cumbersome at times to deal with (atleast for someone coming from pytorch/numpy world), but it has it's own benefits in terms of JIT compilation.

This PR also includes the alias configuration setup required to de-duplicate code with a common d2l syntax. 

Note: JAX always runs everything on GPUs by default if it detects a GPU device, this might not play well with [how d2lbook detects](https://github.com/d2l-ai/d2l-book/blob/c850397e738045f51204dd42ae737e37883ef54a/d2lbook/resource.py#L30) which notebooks to run on GPUs and which not. We may need to patch that, if that becomes an issue in the future.

### CODE REVIEW

File: chapter_linear-regression/linear-regression.md [hunk 1]
```diff
@@ -451,9 +460,22 @@ for i in range(n):
 f'{time.time() - t:.5f} sec'
 ```
 
+```{.python .input}
+%%tab jax
+# JAX arrays are immutable, meaning that once
+# created their contents cannot be changed
+# For updating individual elements, JAX provides
+# an indexed update syntax that returns an updated copy
```
Comment (astonzhang) [inline chapter_linear-regression/linear-regression.md:468]: Make each line as close to 78 characters as possible before a newline

File: chapter_linear-regression/linear-regression.md [hunk 2]
```diff
@@ -518,10 +543,14 @@ d2l.plot(x.asnumpy(), [normal(x, mu, sigma).asnumpy() for mu, sigma in params],
...
 ```
 
-```{.python .input  n=8}
-%%tab pytorch, tensorflow
-# Use numpy again for visualization
-x = np.arange(-7, 7, 0.01)
+```{.python .input}
+%%tab pytorch, tensorflow, jax
+if tab.selected('jax'):
+    # Use jax numpy for visualization
```
Comment (astonzhang) [inline chapter_linear-regression/linear-regression.md:549]: `JAX NumPy`

File: chapter_linear-regression/linear-regression.md [hunk 3]
```diff
@@ -518,10 +543,14 @@ d2l.plot(x.asnumpy(), [normal(x, mu, sigma).asnumpy() for mu, sigma in params],
...
+    x = jnp.arange(-7, 7, 0.01)
+if tab.selected('pytorch', 'mxnet', 'tensorflow'):
+    # Use numpy again for visualization
```
Comment (astonzhang) [inline chapter_linear-regression/linear-regression.md:552]: NumPy

File: chapter_linear-regression/oo-design.md [hunk 4]
```diff
@@ -66,6 +66,18 @@ from d2l import torch as d2l
...
+```{.python .input}
+%%tab jax
+from dataclasses import field
+import time
+import numpy as np
+from d2l import jax as d2l
+import jax
+import jax.numpy as jnp
+from flax import linen as nn
+from flax.training.train_state import TrainState
```
Comment (astonzhang) [inline chapter_linear-regression/oo-design.md:78]: order alphabetically

File: chapter_linear-regression/oo-design.md [hunk 5]
```diff
@@ -162,26 +174,49 @@ Sometimes we put the code to compute the output into a separate `forward` method
...
+    if tab.selected('pytorch', 'mxnet', 'tensorflow'):
+        def __init__(self, plot_train_per_epoch=2, plot_valid_per_epoch=1):
+            super().__init__()
+            self.save_hyperparameters()
+            self.board = ProgressBoard()
         if tab.selected('tensorflow'):
             self.training = None
 
+    if tab.selected('jax'):
+        # No need for save_hyperparam when using python dataclass
```
Comment (astonzhang) [inline chapter_linear-regression/oo-design.md:186]: Python

File: chapter_linear-regression/oo-design.md [hunk 6]
```diff
@@ -162,26 +174,49 @@ Sometimes we put the code to compute the output into a separate `forward` method
...
+        # python dataclasses do not work perfectly well with inheritance
```
Comment (astonzhang) [inline chapter_linear-regression/oo-design.md:187]: Python

File: chapter_linear-regression/oo-design.md [hunk 7]
```diff
@@ -162,26 +174,49 @@ Sometimes we put the code to compute the output into a separate `forward` method
...
     if tab.selected('tensorflow'):
         def call(self, X, *args, **kwargs):
             if kwargs and "training" in kwargs:
                 self.training = kwargs['training']
             return self.forward(X, *args)
 
+    if tab.selected('jax'):
+        # JAX & Flax don't have a forward method like syntax
+        # Flax uses setup and built-in __call__ magic methods for forward pass
+        # Adding here for consistency
```
Comment (astonzhang) [inline chapter_linear-regression/oo-design.md:212]: Group these three lines into one comment with multiple sentences, and apply the greedy rule.

File: chapter_linear-regression/oo-design.md [hunk 8]
```diff
@@ -202,15 +237,31 @@ class Module(d2l.nn_Module, d2l.HyperParameters):  #@save
...
+                                                     batch[-1])
+            self.plot("loss", l, train=True)
+            return l, grads
 
-    def validation_step(self, batch):
-        l = self.loss(self(*batch[:-1]), batch[-1])
-        self.plot('loss', l, train=False)
+        def validation_step(self, params, batch):
+            l = self.loss(params, *batch[:-1], batch[-1])
+            self.plot("loss", l, train=False)
```
Comment (astonzhang) [inline chapter_linear-regression/oo-design.md:264]: single quotes

File: chapter_linear-regression/synthetic-regression-data.md [hunk 9]
```diff
@@ -42,6 +42,16 @@ import tensorflow as tf
 import random
 ```
 
+```{.python .input}
+%%tab jax
+%matplotlib inline
+from d2l import jax as d2l
+import jax
+import jax.numpy as jnp
```
Comment (astonzhang) [inline chapter_linear-regression/synthetic-regression-data.md:50]: from xx import yy

File: chapter_linear-regression/synthetic-regression-data.md [hunk 10]
```diff
@@ -173,6 +188,15 @@ As before, we need a dataset with features `X` and labels `y`.
 Beyond that, we set `batch_size` in the built-in data loader 
 and let it take care of shuffling examples  efficiently.
 
+:begin_tab:`jax`
+JAX is laser-focused on program transformations and
+accelerator-backed NumPy, so it doesn’t include data loading or munging
+in the JAX library. There are already a lot of great data loaders out
+there, and jax suggests using them instead of reinventing anything.
```
Comment (astonzhang) [inline chapter_linear-regression/synthetic-regression-data.md:195]: JAX

File: chapter_linear-regression/synthetic-regression-data.md [hunk 11]
```diff
@@ -173,6 +188,15 @@ As before, we need a dataset with features `X` and labels `y`.
...
+Here we'll grab PyTorch’s data loader, and modify it slightly to make
```
Comment (astonzhang) [inline chapter_linear-regression/synthetic-regression-data.md:196]: we will

File: chapter_linear-regression/synthetic-regression-data.md [hunk 12]
```diff
@@ -173,6 +188,15 @@ As before, we need a dataset with features `X` and labels `y`.
 Beyond that, we set `batch_size` in the built-in data loader 
 and let it take care of shuffling examples  efficiently.
 
+:begin_tab:`jax`
+JAX is laser-focused on program transformations and
+accelerator-backed NumPy, so it doesn’t include data loading or munging
```
Comment (astonzhang) [inline chapter_linear-regression/synthetic-regression-data.md:193]: so the current version (v0.XXX) does not include data loading or ...

File: chapter_linear-regression/synthetic-regression-data.md [hunk 13]
```diff
@@ -186,6 +210,35 @@ def get_tensorloader(self, tensors, train, indices=slice(0, None)):
...
                                            shuffle=train)
+    if tab.selected('jax'):
+        # Use PyTorch Dataset and Dataloader
+        # JAX or Flax do not provide any dataloading functionality
+        from torch.utils import data
+
+        def jax_collate(batch):
+            if isinstance(batch[0], np.ndarray):
+                return jnp.stack(batch)
+            elif isinstance(batch[0], (tuple,list)):
```
Comment (astonzhang) [inline chapter_linear-regression/synthetic-regression-data.md:221]: add a space before list

File: chapter_linear-regression/linear-regression-scratch.md [hunk 14]
```diff
@@ -89,6 +99,19 @@ class LinearRegressionScratch(d2l.Module):  #@save
...
+class LinearRegressionScratch(d2l.Module):  # @save
+    num_inputs: int
+    lr: float
+    sigma: float = 0.01
+
+    def setup(self):
+        self.w = self.param('w', nn.initializers.normal(self.sigma),
+                            (self.num_inputs, 1))
+        self.b = self.param('b', nn.initializers.zeros, (1))
+```
```
Comment (astonzhang) [inline chapter_linear-regression/linear-regression-scratch.md:113]: merge this into the above code block by using `if tab.selected('jax'):`. Otherwise there's no point to only merge the other three framework implementations and keep jax separate

File: chapter_linear-regression/linear-regression-scratch.md [hunk 15]
```diff
@@ -89,6 +99,19 @@ class LinearRegressionScratch(d2l.Module):  #@save
             self.b = tf.Variable(b, trainable=True)
 ```
 
+```{.python .input}
+%%tab jax
+class LinearRegressionScratch(d2l.Module):  # @save
+    num_inputs: int
+    lr: float
+    sigma: float = 0.01
```
Comment (astonzhang) [inline chapter_linear-regression/linear-regression-scratch.md:107]: Is this the first occurrence of such practices? If yes, can you add more explanations in text?

File: chapter_linear-regression/linear-regression-scratch.md [hunk 16]
```diff
@@ -215,27 +247,57 @@ class SGD(d2l.HyperParameters):  #@save
...
 
+```{.python .input  n=8}
+%%tab jax
+class SGD(d2l.HyperParameters):  #@save
+    def __init__(self, lr):
+        """
+        Minibatch stochastic gradient descent.
+        The key transformation of Optax is the `GradientTransformation`
+        defined by two functions, the `init` and the `update`.
+        The `init` initializes the state and the `update` transforms
```
Comment (astonzhang) [inline chapter_linear-regression/linear-regression-scratch.md:270]: Remove ` in code comments (we have updated the style guide to remove them to keep comments shorter)

File: chapter_linear-regression/linear-regression-scratch.md [hunk 17]
```diff
@@ -215,27 +247,57 @@ class SGD(d2l.HyperParameters):  #@save
...
 ```
 
+```{.python .input  n=8}
+%%tab jax
+class SGD(d2l.HyperParameters):  #@save
+    def __init__(self, lr):
+        """
+        Minibatch stochastic gradient descent.
+        The key transformation of Optax is the `GradientTransformation`
+        defined by two functions, the `init` and the `update`.
```
Comment (astonzhang) [inline chapter_linear-regression/linear-regression-scratch.md:269]: functions -> methods

File: chapter_linear-regression/linear-regression-scratch.md [hunk 18]
```diff
@@ -215,27 +247,57 @@ class SGD(d2l.HyperParameters):  #@save
...
+        The `init` initializes the state and the `update` transforms
+        the gradients.
+        https://github.com/deepmind/optax/blob/master/optax/_src/transform.py
+        """
+        self.save_hyperparameters()
+
+    def init(self, params):
+        # params not used
```
Comment (astonzhang) [inline chapter_linear-regression/linear-regression-scratch.md:277]: Use complete sentences (with verb): Delete unused params

File: chapter_linear-regression/linear-regression-scratch.md [hunk 19]
```diff
@@ -215,27 +247,57 @@ class SGD(d2l.HyperParameters):  #@save
...
+        del params
+        return optax.EmptyState
+
+    def update(self, updates, state, params=None):
+        del params
+        # apply_gradients through flax's train_state is called, which then
+        # calls optax.apply_updates adding the params to the update defined
```
Comment (astonzhang) [inline chapter_linear-regression/linear-regression-scratch.md:284]: Not very clear. Can you rephrase?

File: chapter_linear-regression/weight-decay.md [hunk 20]
```diff
@@ -413,6 +442,18 @@ class WeightDecay(d2l.LinearRegression):
         return super().loss(y_hat, y) + self.net.losses
 ```
 
+```{.python .input}
+%%tab jax
+class WeightDecay(d2l.LinearRegression):
+    wd: int = 0
+    
+    def configure_optimizers(self):
+        # weight_decay is not available within `optax.sgd`, but
```
Comment (astonzhang) [inline chapter_linear-regression/weight-decay.md:451]: What is "weight_decay"? A variable name or a method name?

File: chapter_linear-regression/oo-design.md [hunk 21]
```diff
@@ -281,15 +332,42 @@ class Trainer(d2l.HyperParameters):  #@save
...
+                self.key = jax.random.PRNGKey(0)  # Avoid, but use as fallback
+            input_shape = next(iter(self.train_dataloader))[0].shape
+            dummy_input = jnp.zeros(input_shape)
+            params = self.model.init(self.key, dummy_input)
+            return params
+
+        def fit(self, model, data, key=None):
+            self.prepare_data(data)
+            self.prepare_model(model)
+            self.params = self.apply_init(key=key)
```
Comment (astonzhang) [inline chapter_linear-regression/oo-design.md:360]: Consider defining a d2l.SEED constant to be fed into key=jax.random.PRNGKey(d2l.SEED) here. So we can simplify usage of this method like:

```
key = jax.random.PRNGKey(24)
trainer.fit(model, data, key)
```
as
```
trainer.fit(model, data)
```

File: chapter_linear-regression/oo-design.md [hunk 22]
```diff
@@ -162,26 +174,49 @@ Sometimes we put the code to compute the output into a separate `forward` method
...
+            return self.net(X)
 
     if tab.selected('tensorflow'):
         def call(self, X, *args, **kwargs):
             if kwargs and "training" in kwargs:
                 self.training = kwargs['training']
             return self.forward(X, *args)
 
+    if tab.selected('jax'):
+        # JAX & Flax don't have a forward method like syntax
```
Comment (astonzhang) [inline chapter_linear-regression/oo-design.md:210]: forward-method-like

**Review by astonzhang — COMMENTED** [review]
For the following issues, I only added comments to selected places and they may apply to elsewhere. Please scan the entire chapter to fix:

* Readers should not be encouraged to pay their attention to the change of random seeds in different sections. Consider simplifying
```
key = jax.random.PRNGKey(24) or jax.random.PRNGKey(1), or jax.random.PRNGKey(A_DIFFERENT_INT)
trainer.fit(model, data, key)
```
as
```
trainer.fit(model, data)
```
* Treat code comments as sentences (or noun phrases when illustrating variables)
* In code (including comments), be greedy in making newline: keep the current line as close to 78 chars as possible before making newline
* In code comments, add a period for a complete sentence if it's not the end of the current comment
* Sort import statement alphabetically
* In text/code comments, python -> Python, jax -> JAX, numpy -> NumPy, flax -> Flax, optax -> Optax, etc.
* Distinguish methods and functions

File: chapter_linear-regression/linear-regression-concise.md [hunk 23]
```diff
@@ -60,6 +60,16 @@ import numpy as np
...
 
+```{.python .input  n=1}
+%%tab jax
+from d2l import jax as d2l
+import jax
+from jax import numpy as jnp
+from jax import grad
+from flax import linen as nn
+import optax
+```
```
[the remark this answers is not in the document]
Comment (AnirudhDagar) [reply chapter_linear-regression/linear-regression-concise.md:71]: Is that something we do everywhere? Because alphabetical imports are inconsistent across d2l even outside of JAX. We should then fix everything.

Also, I thought it didn't matter. We use `isort` to format imports anyway https://github.com/d2l-ai/d2l-book/pull/27/commits/96ab562095f1d613700213625e7abee4f164e949

File: chapter_linear-regression/oo-design.md [hunk 24]
```diff
@@ -281,15 +332,42 @@ class Trainer(d2l.HyperParameters):  #@save
...
+            self.val_batch_idx = 0
+            for self.epoch in range(self.max_epochs):
+                self.fit_epoch()
+
+    if tab.selected('jax'):
+        def apply_init(self, **kwargs):
+            if kwargs and 'key' in kwargs and (kwargs['key'] is not None):
+                self.key = kwargs['key']
+            else:
+                self.key = jax.random.PRNGKey(0)  # Avoid, but use as fallback
```
[the remark this answers is not in the document]
Comment (AnirudhDagar) [reply chapter_linear-regression/oo-design.md:350]: This is an intentional design choice I made for JAX. Readers SHOULD pay attention to setting up seed in JAX.
Why? Please read this -> https://jax.readthedocs.io/en/latest/jax.random.html#prng-keys

> Unlike the stateful pseudorandom number generators (PRNGs) that users of NumPy and SciPy may be accustomed to, JAX random functions all require an explicit PRNG state to be passed as a first argument.

---
2nd part

> It's also not user-friendly to change seeds in different sections (readers should not be encouraged to pay their attention to the change of seed)

If the seed is not changed, then the results will be the same always (still are). For all other frameworks, we don't use any SEED. I remember bringing this discussion in D2L slack two years ago and we decided to set no seeds anywhere.
Infact i would suggest setting the SEED also randomly -> key = jax.random.PRNGKey(random.randint(0,100)) to make our implementation more robust.

---
My personal opinion:

As we wrap away more and more things, we are making things no different for any framework. What will be the point of having different frameworks if we don't highlight their differences? Setting up PRNGs is an important step, and hence I added that as a param here.

File: chapter_linear-regression/oo-design.md (hunk 21 above)
Comment (AnirudhDagar) [reply chapter_linear-regression/oo-design.md:359]: See comment above.

File: chapter_linear-regression/synthetic-regression-data.md (hunk 12 above)
Comment (AnirudhDagar) [reply chapter_linear-regression/synthetic-regression-data.md:193]: This is straight out of their documentation. See [here](https://jax.readthedocs.io/en/latest/notebooks/Neural_Network_and_Data_Loading.html#data-loading-with-pytorch). I don't think JAX will ever have a dataloader. Even if they do get something, it will be an external package in the JAX Ecosystem.

Hardcoding a version here is also not a wise choice IMO

File: chapter_linear-regression/synthetic-regression-data.md (hunk 13 above)
Comment (AnirudhDagar) [reply chapter_linear-regression/synthetic-regression-data.md]: Sorry missed that! 😬

File: chapter_linear-regression/weight-decay.md (hunk 20 above)
Comment (AnirudhDagar) [reply chapter_linear-regression/weight-decay.md]: Ahh, it is no name, I'll rephrase this.

**Review by AnirudhDagar — COMMENTED** [review]
I've adddressed most of the style changes and made some comments on a few design choices.

File: chapter_linear-regression/oo-design.md (hunk 5 above)
Comment (astonzhang) [reply chapter_linear-regression/oo-design.md]: Add a period to the end of a sentence if this sentence is not the end of the comment. Don't add multiple comments for the same line of code: just merge them into a paragraph of multiple sentences.

File: chapter_linear-regression/linear-regression-concise.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (astonzhang) [reply chapter_linear-regression/linear-regression-concise.md:71]: Yes that's for backup in the final presentation. A benefit of sorting them in each markdown file is that it's easy to avoid duplications

[the remark this answers is not in the document]
Comment (AnirudhDagar) [reply chapter_linear-regression/linear-regression-concise.md:71]: cool, I'll update that. let's be consistent about that everywhere then. There tonnes of code blocks which are not following the alphabetical import style

File: chapter_linear-regression/oo-design.md (hunk 5 above)
Comment (AnirudhDagar) [reply chapter_linear-regression/oo-design.md]: Cool, then we can just leave this as is because I removed the python dataclass comment in the last commit since it was confusing for the readers.

File: chapter_linear-regression/linear-regression-scratch.md (hunk 15 above)
Comment (astonzhang) [reply chapter_linear-regression/linear-regression-scratch.md:107]: To be more beginner-friendly, can you describe it briefly in OOD section?

File: chapter_linear-regression/linear-regression-scratch.md [hunk 25]
```diff
@@ -89,6 +99,19 @@ class LinearRegressionScratch(d2l.Module):  #@save
             self.b = tf.Variable(b, trainable=True)
 ```
 
+```{.python .input}
+%%tab jax
+class LinearRegressionScratch(d2l.Module):  # @save
```
Comment (astonzhang) [inline chapter_linear-regression/linear-regression-scratch.md:104]: Remove the space before @save

File: chapter_linear-regression/synthetic-regression-data.md [hunk 26]
```diff
@@ -189,12 +189,11 @@ Beyond that, we set `batch_size` in the built-in data loader
...
 
 :begin_tab:`jax`
-JAX is laser-focused on program transformations and
-accelerator-backed NumPy, so it doesn’t include data loading or munging
-in the JAX library. There are already a lot of great data loaders out
-there, and jax suggests using them instead of reinventing anything.
-Here we'll grab PyTorch’s data loader, and modify it slightly to make
-it work with NumPy arrays.
+JAX is all about NumPy like API with device acceleration and the functional
+transformations, so atleast the current version doesn’t include data loading
```
Comment (astonzhang) [inline chapter_linear-regression/synthetic-regression-data.md:193]: at least

File: chapter_linear-regression/synthetic-regression-data.md [hunk 27]
```diff
@@ -189,12 +189,11 @@ Beyond that, we set `batch_size` in the built-in data loader
...
+methods. With other  libraries we already have great data loaders out there,
+and JAX suggests using them instead. Here we will grab PyTorch’s data loader,
+and modify it slightly to make it work with JAX.
```
Comment (astonzhang) [inline chapter_linear-regression/synthetic-regression-data.md:196]: "Here we will grab PyTorch’s data loader, and modify it slightly to make it work with JAX."

Are we using/modifying PyTorch's data loader here?

Comment (AnirudhDagar) [reply chapter_linear-regression/synthetic-regression-data.md:196]: Yes, we use a custom `collate_fn` and a custom dataset class for JAX which is built on top of pytorch's dataset and dataloader.

File: chapter_linear-regression/synthetic-regression-data.md [hunk 28]
```diff
@@ -186,6 +209,35 @@ def get_tensorloader(self, tensors, train, indices=slice(0, None)):
         dataset = torch.utils.data.TensorDataset(*tensors)
         return torch.utils.data.DataLoader(dataset, self.batch_size,
                                            shuffle=train)
+    if tab.selected('jax'):
+        # Use PyTorch Dataset and Dataloader
+        # JAX or Flax do not provide any dataloading functionality
+        from torch.utils import data
```
Comment (astonzhang) [inline chapter_linear-regression/synthetic-regression-data.md:215]: `data` is too common and it may conflict with others. So let's do `import torch` and `torch.utils.data.Dataset` (see https://github.com/d2l-ai/d2l-en/blob/master/d2l/torch.py#L244). Besides, we add import statement either at the Preface section (to be saved in d2l libraries) or at the beginning of each section.

Comment (AnirudhDagar) [reply chapter_linear-regression/synthetic-regression-data.md:215]: Yes I plan to get rid of this, earlier the idea was to isolate all the pytorch imports only in the dataset functions so that readers aren't confused. But that will not be possible, we will have to import torch and torchvision. I'll send a PR to fix this.
